### DISCUSSION TITLE
Neue Features

### DISCUSSION THREAD

**Discussion by drakesoft** [shining-man/bsc_fw#65] [Ideas]
Hallo,
ich benötige noch folgende Funktionen:
1. virtuelle "Digital Eingänge" mit input von MQTT
2. SDM630 Zähler emulation mit Input von MQTT
Hintergrund ist folgender: Meine Zähler lese ich mit esphome aus und möchte gerne die Daten weiter reichen an meinen Deye Hybrid Inverter - der erwartet einen SDM630 am RS485 mit Modbus. Die virtuellen Digitaleeingänge benötige ich um einen Ausgang zu schalten um den Wechselrichter sofern er nicht benötigt wird abzuschalten.
@shining-man Ich würde das ganze selbst Programmieren und den Code wieder in dein Projekt integrieren, falls gewünscht. Bitte um Input.

Danke und frohe Weihnachen

drakesoft

**Comment by shining-man** [comment]
Zu 1.: Das kann ich einbauen, da haben schon mehr danach gefragt. Hier wäre noch die Frage was besser wäre MQTT oder Restapi?
Zu 2.: Wenn ich es richtig versthe, willst du einen Wert, den du per MQTT bekommst, an den DEYE via RS485 weiterleiten. Bisher habe ich immer nur Werte von einem RS485 Device geholt, aber noch nichts hingesendet. Ich bin zwar gerade dabei, diese Route einzubauen. Dazu gibt es eine Funktion die sich setSerialBmsWriteData() nennt. Mit der Funktion bekommst du die Daten bis in den serialHanlder (BscSerial.cpp), der es dann weiter an den jeweiligen serialDeviceHanlder (diese findest du unter /devices) gibt. So einen serialDeviceHanlder musst du dir für deinen DEYE schreiben.
Mir gefällt die jetzige Lösung mit dem setSerialBmsWriteData() zwar noch nicht ganz so gut, es ist aber aktuell die einzige Möglichkeit Daten von der Applikation an das serielle Geräte zu bekommen.
Jetzt musst du mir nur genau sagen, wo su Input brauchst?

Eine andere Frage. Warum liest du nicht mit esphome den Zähler aus und lässt auch gleich diesen den Wert an den Wechselrichter senden?
So ist es ein langer Weg: esphome -> mqtt -> BSC -> DEYE
Es ist auch die Frage wie zeitlich kritisch das ganze ist? Denn wenn du mit setSerialBmsWriteData() die Werte zum Senden übergibst, dann wird das erledigt wenn der SerialHandler das nächste mal den Port bearbeiten, an dem dein DEYE angeschlossen ist. Ich bearbeite die Ports alle 1000ms. Je nachdem wieviele BMS du am BSC angeschlossen hast, wird die Zeit dann länger.

**Reply by drakesoft** [reply]
danke für deine schnelle Antwort.
zu 1. ich denke MQTT wäre einfacher. Eigentlich müsste man nur z.B. die Topic /bsc/input/in0 usw.. subscriben. pubsubclient ruft dann bei änderung den hinterlegten Callback auf und man macht ein Update der virtuellen Input Zustände. Das ganze aussen rum mit Connect, Authentifizierung usw.. ist ja schon alles da.
zu 2. klar liese sich das auch mit esphome realisieren. Problem ist nur das ich soviel Zähler auswerten muss das kein Port mehr frei ist. Und am Wechselrichter hängt mit dem Akku der BSC dran wo sogar noch zwei frei wären. 
Ja das Polling der Ports ist dann schon ein Problem. Da man in dem Fall nicht der Master sondern der Slave ist.

**Comment by shining-man** [comment]
zu 1.: Wie das Umzusätzen ist, ist mir klar. In der Umsetzung ist beides einfach. Die Frage ist, was ist aus Sicht des Users besser/schöner. Bei MQTT brauchst du einen Broker am laufen, was du über die Restapi nicht brauchst.
zu 2.: Hier ist für mich immer noch die Frage, wo du genau von mir den Input brauchst?    Vmtl. wäre es einfacher du würdest mit dem BSC die Werte von einem Zähler holen und diese per MQTT weiter zu senden. Dann hast du bei demem esphome Device einen Port frei.
